### PR TITLE
format: move layers, scenarios and steps to bottom of scope.

### PR DESCRIPTION
**Pull request by berniexie**
Extract our board nodes and moves them to the end of the node slice before writing it as a string. Adds a double new line before the root level ones unless they already have one in front of them.

### CODE REVIEW

File: d2ast/d2ast.go [hunk 1]
```diff
@@ -760,8 +760,8 @@ func (mb MapNodeBox) Unbox() MapNode {
 	}
 }
 
-func (mb MapNodeBox) IsSpecialBoard() bool {
-	if mb.MapKey == nil || mb.MapKey.Key == nil {
+func (mb MapNodeBox) IsBoardNode() bool {
+	if mb.MapKey == nil || mb.MapKey.Key == nil || len(mb.MapKey.Key.Path) == 0 {
```
Comment (alixander) [inline d2ast/d2ast.go:798]: != 1

File: d2format/format_test.go [hunk 2]
```diff
@@ -657,6 +657,137 @@ x: @../file
...
+
+b
+
+c
+d
+
+layers: {
+  Test super nested: {
+    base-layer
+
```
Comment (alixander) [inline d2format/format_test.go:755]: hmm there wasn't a newline in between before

File: ci/release/changelogs/next.md [hunk 3]
```diff
@@ -4,9 +4,16 @@
 
 #### Improvements 🧹
 
+- Use shape specific sizing for grid containers [#1294](https://github.com/terrastruct/d2/pull/1294)
```
Comment (alixander) [inline ci/release/changelogs/next.md:7]: .

File: d2format/format.go [hunk 4]
```diff
@@ -302,6 +328,13 @@ func (p *printer) _map(m *d2ast.Map) {
 			p.sb.WriteString("; ")
 		}
 
+		if m.IsFileMap() && nb.IsBoardNode() {
+			currString := p.sb.String()
+			// if the two characters before the board node is not a double newline, we add one
+			if currString[len(currString)-2:] != "\n\n" {
+				p.newline()
+			}
+		}
```
Comment (alixander) [inline d2format/format.go:337]: i don't understand this part

File: d2format/format.go [hunk 5]
```diff
@@ -306,6 +323,25 @@ func (p *printer) _map(m *d2ast.Map) {
...
+	boards := []d2ast.MapNodeBox{}
+	boards = append(boards, layerNodes...)
+	boards = append(boards, scenarioNodes...)
+	boards = append(boards, stepNodes...)
+
+	// draw board nodes
+	for i := 0; i < len(boards); i++ {
+		n := boards[i].Unbox()
+		// if this board is the very first line of the file, don't add an extra indent
+		if n.GetRange().Start.Line != 0 {
```
Comment (alixander) [inline d2format/format.go:335]: check column == 0 too. who knows, maybe someone wants to 1-line everything

File: d2format/format.go [hunk 6]
```diff
@@ -306,6 +323,25 @@ func (p *printer) _map(m *d2ast.Map) {
...
+			p.newline()
+		}
+		if len(m.Nodes) > len(boards) {
+			p.newline()
+		}
```
Comment (alixander) [inline d2format/format.go:340]: can u explain this w/ an example or test?

**Review by alixander — CHANGES_REQUESTED** [review]

File: d2format/format.go (hunk 5 above)
Comment (berniexie) [reply d2format/format.go:335]: sorry, my comment shouldn't and instead should say newline.

**Review by alixander — APPROVED** [review]

**Pull request merged**